File: js/Json/StringTable.js
import { item, map, setItem, fill } from "../fable_modules/fable-library-js.4.22.0/Array.js";
import { iterate } from "../fable_modules/fable-library-js.4.22.0/Seq.js";
import { int, string, array as array_2 } from "../fable_modules/Thoth.Json.Core.0.4.0/Decode.fs.js";
import { Dictionary_tryFind } from "../Core/Helper/Collections.js";
import { addToDict } from "../fable_modules/fable-library-js.4.22.0/MapUtil.js";
import { FSharpResult$2 } from "../fable_modules/fable-library-js.4.22.0/Result.js";

export function arrayFromMap(otm) {
    const a = fill(new Array(otm.size), 0, otm.size, "");
    iterate((kv) => {
        setItem(a, kv[1], kv[0]);
    }, otm);
    return a;
}

export function encoder(ot) {
    const values = map((value) => ({
        Encode(helpers) {
            return helpers.encodeString(value);
        },
    }), ot);
    return {
        Encode(helpers_1) {
            const arg = map((v) => v.Encode(helpers_1), values);
            return helpers_1.encodeArray(arg);
        },
    };
}

export const decoder = array_2(string);

export function encodeString(otm, s) {
    const matchValue = Dictionary_tryFind(s, otm);
    if (matchValue == null) {
        const i_1 = otm.size | 0;
        addToDict(otm, s, i_1);
        return {
            Encode(helpers_1) {
                return helpers_1.encodeSignedIntegralNumber(i_1);
            },
        };
    }
    else {
        const i = matchValue | 0;
        return {
            Encode(helpers) {
                return helpers.encodeSignedIntegralNumber(i);
            },
        };
    }
}

export function decodeString(ot) {
    return {
        Decode(s, json) {
            const matchValue = int.Decode(s, json);
            return (matchValue.tag === 1) ? (new FSharpResult$2(1, [matchValue.fields[0]])) : (new FSharpResult$2(0, [item(matchValue.fields[0], ot)]));
        },
    };
}

File: src/Json/StringTable.fs
namespace rec ARCtrl.Json

open Thoth.Json.Core

open ARCtrl

open ARCtrl.Helper

module StringTable =

    type StringTableMap = System.Collections.Generic.Dictionary<string,int>

    type StringTableArray = array<string>

    let arrayFromMap (otm : StringTableMap) : StringTableArray =
        let a = Array.zeroCreate<string> otm.Count
        otm
        |> Seq.iter (fun kv -> a.[kv.Value] <- kv.Key)
        a

    let encoder (ot: StringTableArray) =
        ot
        |> Array.map Encode.string
        |> Encode.array

    let decoder : Decoder<StringTableArray> =
        Decode.array Decode.string
        
    let encodeString (otm : StringTableMap) (s : string) =
        match StringDictionary.tryFind s otm with
        | Some i -> Encode.int i
        | None ->
            let i = otm.Count
            otm.Add(s,i)
            Encode.int i

    let decodeString (ot : StringTableArray) : Decoder<string> = 
        { new Decoder<string> with
            member this.Decode (s,json) = 
                match Decode.int.Decode(s,json) with
                | Ok i -> Ok ot.[i]
                | Error err -> Error err
        }
